### PR TITLE
[SW] Add kernels: gemv, spmv, conjugate gradient

### PR DESCRIPTION
**Pull request by husterZC**
Description of PR that completes issue here...

### Added

- spmv
- gemv
- conjugate gradient

### CODE REVIEW

**Comment by mp-17** [comment]
Thanks a lot @husterZC, I will check it asap!

**Comment by husterZC** [comment]
Hi Matteo, one thing to notice: 

In the conjugate_gradient/shared_kernels. these are all soft link files, but I don't know why we I push them, they loss the link. Could you please fix them yourself, either make a hard copy or make a link file.  And for some workloads using other kernels, do you suggest to make a link or a hard copy?

**Comment by mp-17** [comment]
I would prefer symlinks to avoid duplicated code. I created a PR to your `main`: `https://github.com/husterZC/ara/pull/1`

File: apps/common/default_args.mk [hunk 1]
```diff
@@ -26,3 +26,7 @@ def_args_softmax     = "3 256"
 def_args_pathfinder  = "1 1024 64"
 # Batch_size, depth, height, width, n_boxes (in total), crop_h, crop_w
 def_args_roi_align   = "1 32 4 4 4 2 2"
+#SpMV configuration: row, col, density
+def_args_spmv				 = "128 128 0.6"
```
Comment (mp-17) [inline apps/common/default_args.mk:30]: Please adjust vertical alignment ^

File: apps/conjugate_gradient/main.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+extern double r[] __attribute__((aligned(4 * NR_LANES), section(".l2")));
+extern double p[] __attribute__((aligned(4 * NR_LANES), section(".l2")));
+extern double Ap[] __attribute__((aligned(4 * NR_LANES), section(".l2")));
+extern int32_t A_PROW[] __attribute__((aligned(4 * NR_LANES), section(".l2")));
+extern int32_t A_IDX[] __attribute__((aligned(4 * NR_LANES), section(".l2")));
+extern double A_DATA[] __attribute__((aligned(4 * NR_LANES), section(".l2")));
+
+void daxpy(double * x, double a, double * y, double * dest, uint64_t len){
+	#define MAX_AVL 128
+	while(len>MAX_AVL){
```
Comment (mp-17) [inline apps/conjugate_gradient/main.c:56]: Have you passed this through `clang-format`? 
To do so, just run `make -C apps format`

**Review by mp-17 — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution! 
Can you please pass everything through `clang-format` and adjust the vertical alignment in `apps/common/default_args.mk`?
Then, if the CI checks pass, LGTM!

**Comment by mp-17** [comment]
@husterZC Ping me if you need support! :-)

**Comment by husterZC** [comment]
To be honest, I'm busy on other stuff. I plan to fix this last in this week

**Comment by husterZC** [comment]
Hey Matteo,

I've added the dependence of `sklearn` python package for conjugate gradient kernel.
We need this one to generate the symmetric and positive-defined (SPD) dense/sparse matrix for conjugate gradient
If you find it is a bit troublesome to add new dependency, there are other solutions:
1. I can bring a pre-build SPD matrix and delete the function to generate SPD matrix at given size and sparsity. Well that  would be very boring, and not very meaningful.
2. I can look into how to generate SPD matrix at given size and sparsity, drive myself to be a mathematician and take somedays of working.

**Comment by mp-17** [comment]
https://github.com/pulp-platform/ara/actions/runs/3563364133/jobs/5989785427#step:8:289

**Comment by husterZC** [comment]
Hi Matteo, I wonder how to add the python package "sklearn" in the CI environment, otherwise it will always sticking here ;)

**Review by mp-17 — APPROVED** [review]

**Comment by mp-17** [comment]
I will check everything and merge here -> https://github.com/pulp-platform/ara/pull/267
Thanks again Chi!

**Pull request closed without merging**